CRAMcodecs.tex

```diff
@@ -0,0 +1,3183 @@
+frequencies are normalised to add up to 4095.
+
+Formally, this is an ordered alphabet $\mathbb{A}$ containing symbols $s$ where
+$s_{i}$ with the $i$-th symbol in $\mathbb{A}$, occurring with the frequency $freq_{i}$.
+
+\subsubsection*{Order-0 encoding}
+
+The normalised symbol frequencies are then written out as \{symbol,
+frequency\} pairs in ascending order of symbol (0 to 255 inclusive).
+If a symbol has a frequency of 0 then it is omitted.
+
+To avoid storing long consecutive runs of symbols if all are present
+(eg a-z in a long piece of English text) we use run-length-encoding on
+the alphabet symbols.  If two consecutive symbols have non-zero
+frequencies then a counter of how many other non-zero frequency
+consecutive symbols is output directly after the second consecutive
+symbol, with that many symbols being subsequently omitted.
+
+For example for non-zero frequency symbols `a', `b', `c', `d' and `e'
+we would write out symbol `a', `b' and the value 3 (to indicate `c',
+`d' and `e' are also present).
+
+The frequency is output after every symbol (whether explicit or
+implicit) using ITF8 encoding. This means that frequencies 0-127 are
+encoded in 1 byte while frequencies 128-4095 are encoded in 2 bytes.
+
+Finally the symbol 0 is written out to indicate the end of the
+symbol-frequency table.
+
+As an example, take the string \texttt{abracadabra}.
+
+\begin{minipage}[t]{0.5\textwidth}
+Symbol frequency:
+\\[8pt]
+\begin{tabular}{ |r|r| }
+\hline
+Symbol & Frequency\\
+\hline
+a & 5 \\
+b & 2 \\
+c & 1 \\
+d & 1 \\
+r & 2 \\
+\hline
+\end{tabular}
+\end{minipage}
+\begin{minipage}[t]{0.5\textwidth}
+Normalised to sum to 4095:
+\\[8pt]
+\begin{tabular}{ |r|r|}
+\hline
+Symbol & Frequency\\
+\hline
+a & 1863 \\
+b &  744 \\
+c &  372 \\
+d &  372 \\
+r &  744 \\
+\hline
+\end{tabular}
+\end{minipage}
+
+Encoded as:
+\begin{verbatim}
+0x61      0x87 0x47      # `a'           <1863>
+0x62 0x02 0x82 0xe8      # `b' <+2: c,d>  <744>
+          0x81 0x74      # `c' (implicit) <372>
+          0x81 0x74      # `d' (implicit) <372>
+0x72      0x82 0xe8      # `r'            <744>
+0x00                     # <0>
+\end{verbatim}
+
+
+\subsubsection*{Order-1 encoding}
+
+To encode Order-1 statistics typically requires a larger table as for
+an $N$ sized alphabet we need to potentially store an $N$x$N$ matrix.
+We store these as a series of Order-0 tables.
+
+We start with the outer context byte, emitting the symbol if it is
+non-zero frequency.  We perform the same run-length-encoding as we
+use for the Order-0 table and end the contexts with a nul byte.  After
+each context byte we emit the Order-0 table relating to that context.
+
+One last caveat is that we have no context for the first byte in the
+data stream (in fact for 4 equally spaced starting points, see
+``interleaving" below).  We use the ASCII value (`\textbackslash0') as
+the starting context and so need to consider this in our frequency
+table.
```

Comment:
The output of the rANS implementations in htscodecs also includes the final context, which is neither described here nor in the following example. (The final context in the case of the following example is `('a', '\0')`).

Reply:
Initially I didn't get this -  I dumped out the contents of `out_buf` holding the O1 frequency table for the example input and it matches (bar the frequencies themselves, which don't due to changes in the frequency renormalisation algorithm, but as stated before how you get the frequencies to add up to 4095 is up to the implementation).

However, if I change it to `abracadabraabracadabraabracadabraabracadabrx` so the last letter differs from the first letter, then I see that initially 'x' isn't in the `F[256][256]` frequency table:

```
(gdb) p F['r']
$56 = {0 <repeats 97 times>, 7, 0 <repeats 22 times>, 1, 0 <repeats 135 times>}
(gdb) p F['r']['a']
$57 = 7
(gdb) p F['r']['x']
$58 = 1
(gdb) p F['x'][0]
$59 = 0
(gdb) p F['x']   
$60 = {0 <repeats 256 times>}
```

However F['x'][0..255] gets renormalised because `T['x'] != 0`.  We could work around this special case where the last character in the buffer appears once and once only, meaning it doesn't have any following contexts of its own, with this minor patch:

```
diff --git a/htscodecs/rANS_static.c b/htscodecs/rANS_static.c
index 9865022..9a37eb0 100644
--- a/htscodecs/rANS_static.c
+++ b/htscodecs/rANS_static.c
@@ -420,6 +420,12 @@ unsigned char *rans_compress_O1(unsigned char *in, unsigned int in_size,
         if (T[i] == 0)
             continue;
 
+        if (T[i] == 1 && in[in_size-1] == i) {
+            // last symbol can be ignored as it doesn't have following
+            // contexts.
+            continue;
+        }
+
         //uint64_t p = (TOTFREQ * TOTFREQ) / t;
         double p = ((double)TOTFREQ)/T[i];
     normalise_harder:
```

It changes the output as expected:

```
@ seq4c[samtools.../htscodecs]; hd /tmp/_o1.orig 
00000000  01 48 00 00 00 01 3f 00  00 00 2c 00 00 00 00 61  |.H....?...,....a|
00000010  8f ff 00 61 61 82 86 62  02 86 bd 83 5e 83 5e 00  |...aa..b....^.^.|
00000020  62 02 72 8f ff 00 61 8f  ff 00 61 8f ff 00 72 61  |b.r...a...a...ra|
00000030  8d ff 78 82 00 00 78 00  8f ff 00 00 47 98 48 53  |..x...x.....G.HS|
00000040  47 98 48 53 47 98 48 53  34 d7 46 02 ad           |G.HSG.HS4.F..|
0000004d
@ seq4c[samtools.../htscodecs]; hd /tmp/_o1.new
00000000  01 43 00 00 00 01 3a 00  00 00 2c 00 00 00 00 61  |.C....:...,....a|
00000010  8f ff 00 61 61 82 86 62  02 86 bd 83 5e 83 5e 00  |...aa..b....^.^.|
00000020  62 02 72 8f ff 00 61 8f  ff 00 61 8f ff 00 72 61  |b.r...a...a...ra|
00000030  8d ff 78 82 00 00 00 47  98 48 53 47 98 48 53 47  |..x....G.HSG.HSG|
00000040  98 48 53 34 d7 46 02 ad                           |.HS4.F..|
00000048
```

It's subtle, but there are two 'x's in the o1.orig file, with the first showing r->x and the second being x->nul.

Note however this second transition is not *necessary* for decode and the decoder works fine.  This is an implementation tweak (maybe bug), but not something that the specification should require.  Basically we can add more entries to the frequency table if we really wish and it'll have no bearing on the decoding of the data as they're unused.  I'm not sure how best to explain that in the formal specification.